Show count of visible beatmaps at song select

peppy opened the pull request:
- Closes https://github.com/ppy/osu/issues/13837

Brought across from stable. I was going to make this show more relevant information on why things are filtered, but it turns out to be a bit of a PITA. Having the count along is still beneficial I believe.

https://user-images.githubusercontent.com/191335/222648607-b22bffcd-4449-41ae-ab4e-53137dd2a64d.mp4

peppy committed:
Show count of visible beatmaps at song select

peppy commented on osu.Game/Localisation/SongSelectStrings.cs:
```diff
@@ -19,6 +19,11 @@ public static class SongSelectStrings
+        /// <summary>
+        /// "{0} beatmaps displayed"
+        /// </summary>
+        public static LocalisableString BeatmapsDisplayed(int arg0) => new TranslatableString(getKey(@"beatmaps_displayed"), @"{0:#,0} beatmaps displayed", arg0);
```
This is the first instance I can find of having string format specs inside localisable strings. Discussing with @smoogipoo this seems okay (allows translators to adjust as required for each language) but worth nothing and scrutinising.

bdach replied:
The format specs bother me less than the fact that this string is not correctly translatable anyway, because it has a numeral in it and therefore requires being able to specify plural forms for grammatically correct translation. As in, if we want to have this localisable, we probably need to resurrect or rewrite https://github.com/ppy/osu-framework/pull/4918 first.

In fact as is, this is wonky even in English ("1 beatmaps displayed").

bdach replied:
I think not localising for now is best. Most of song select is not localised anyway.

peppy commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -49,13 +49,23 @@ public partial class BeatmapCarousel : CompositeDrawable, IKeyBindingHandler<Glo
+        /// <summary>
+        /// Triggered after filter conditions have finished being applied to the model hierarchy.
+        /// </summary>
+        public Action? FilterApplied;
```
There was no way previously to hook an event to after a filter operation had finished.

peppy commented on osu.Game/Screens/Select/BeatmapCarousel.cs:
```diff
@@ -49,13 +49,23 @@ public partial class BeatmapCarousel : CompositeDrawable, IKeyBindingHandler<Glo
+        /// <summary>
+        /// The total count of non-filtered beatmaps displayed.
+        /// </summary>
+        public int CountDisplayed => beatmapSets.Where(s => !s.Filtered.Value).Sum(s => s.Beatmaps.Count(b => !b.Filtered.Value));
```
I'd like to see this done as part of the filter process, rather than a post-execution linq. But I don't want to refactor the filter process too much just for this change.

peppy reviewed (commented):
Reviewer notes

peppy committed:
Don't localise beatmap count string for now

Joehuu commented:
I think the text should be at the left, see [design rework](https://www.figma.com/file/DXKwqZhD5yyb1igc3mKo1P/Song-Select-Draft-2?node-id=0-1&t=KFHtYduDFI5QduY2-0):

![image](https://user-images.githubusercontent.com/35318437/225534851-e0bae167-b4a0-4314-a8b9-1b1551ef527b.png)

Users will look at the left as that is where the textbox starts. Stable also puts it there.

peppy commented:
Didn't realise this had been added to the new design. Will match it roughly 👍 .

peppy committed:
Move matches string inside text box

peppy commented:
I've also moved the text inside the textbox. Not 100% sure on this because in the current implementation it shows a count regardless of whether a search is present or not, but let's see how it plays out.

Joehuu commented:
Since you incorporated some of the new design into the old textbox, is it also worth using the already implemented [ShearedSearchTextBox](https://github.com/ppy/osu/blob/master/osu.Game/Graphics/UserInterface/ShearedSearchTextBox.cs)? It's not a drop-in replacement as it inherits `CompositeDrawable` with the text box as a child, so maybe another time?

peppy commented:
I tried and the class was not ready.

bdach committed:
Privatise setter

bdach committed:
Fix importing beatmaps not changing count of visible beatmaps

Reproduction steps:

1. Go to song select
2. Open beatmap listing
3. Import a beatmap that would fit the current filter criteria
4. The count of visible beatmaps does not change

Fixed by updating the count on `BeatmapSetsChanged` too.

bdach commented:
I've made a small code quality change (https://github.com/ppy/osu/pull/22773/commits/695ee39b87644f4bf62a52c41b80805b7b772dff), but more importantly, also fixed an issue wherein importing or updating a beatmap set would not properly update the count of visible beatmaps (https://github.com/ppy/osu/pull/22773/commits/ea8da69263ece422d6119dedc1471a8df4654a1a).

Please check that you're OK with the changes.

peppy commented:
Nice catch, looks good.
